### PR TITLE
[WIP]- Start section 3

### PR DESCRIPTION
**Pull request by txsmith**
and restructure to prevent 'watch' from recursing infinitely

### CODE REVIEW

File: report/sections/3-typescript.tex [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+abstractions that have a conceptually high distance to the semantics of JavaScript.
+The result of this effort means that compiled TypeScript code should be suitable 
+for human consumption. Structural typing provides an excellent bridge from the 
+dynamically typed JavaScript world to the statically typed world of TypeScript.
+In this section we will introduce the most important features of TypeScript and
+how they relate to structural types.
+\\
+\\
+Structural types manifest themselves in TypeScript as \textit{Record Types}.
+In the examples above, we have already encoutered some record types, a set of name-type pairs.
```
Comment (TimvdLippe) [inline report/sections/3-typescript.tex]: encountered

Comment (TimvdLippe) [reply report/sections/3-typescript.tex]: name-type or key-values?

File: report/sections/3-typescript.tex [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+Built directly on top of record types are some of the most important aspects of the language:
+\begin{itemize}
+\item classes
+\item interfaces
+\item union types
+\item intersection types
+\end{itemize}
+3
```
Comment (TimvdLippe) [inline report/sections/3-typescript.tex]: typo?

**Review by TimvdLippe — APPROVED** [review]
Initial version seems fine. I suppose that the "describe the sugar" is still TODO?

File: report/sections/3-typescript.tex (hunk 1 above)
Comment (txsmith) [reply report/sections/3-typescript.tex]: I think 'values' would be confusing because they're not values but types

File: report/sections/1-introduction.tex [hunk 3]
```diff
@@ -0,0 +1,31 @@
+\section{Introduction}
+
+TypeScript is a statically typed programming language that is a superset JavaScript (ECMAScript 2015).
```
Comment (TimvdLippe) [inline report/sections/1-introduction.tex]: We use these terms first here, so lets add links/footnotes to tge official websites here

File: report/sections/2-structural-typing.tex [hunk 4]
```diff
@@ -46,11 +48,11 @@ \section{Structural Typing}
...
-The \texttt{|} operator describes that $NumList$ is the discriminated union of $Cons$ and $Nil$.
+The \texttt{|} operator describes that $NumList$ can either be $Cons$ nil $Nil$.
 Note that $Cons$ and $NumList$ are mutually recursive.
 To check whether we can assign a $NumList$ value to another value, 
 we need to check if their unfoldings are equal.
 Since unfolding a recursive type yields an infinite tree, 
 the type checker has to make sure that this process terminates regardless of the possibly infinite type.
 This approach to type checking is called \textit{equi-recursive} because a recursive type should
-be a subtype of its one step unfolding.
+be a subtype of its one step unfolding.\citep{pierce2002types}
```
Comment (TimvdLippe) [inline report/sections/2-structural-typing.tex]: Citations shouldbe before the dot

File: report/sections/3-typescript.tex [hunk 5]
```diff
@@ -0,0 +1,190 @@
+\section{Structural Typing in Typescript}
+\label{sec:structural-typing-typescript}
+TypeScript was built with structural typing as it's backbone,
```
Comment (TimvdLippe) [inline report/sections/3-typescript.tex]: Its

File: report/sections/3-typescript.tex [hunk 6]
```diff
@@ -0,0 +1,190 @@
+\section{Structural Typing in Typescript}
+\label{sec:structural-typing-typescript}
+TypeScript was built with structural typing as it's backbone, 
+nearly any feature in the language can be described in terms of structual types.
+The reason for this being that the goal of the language is to provide a type system
```
Comment (TimvdLippe) [inline report/sections/3-typescript.tex]: Being,

File: report/sections/3-typescript.tex [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+  size(): number {
+    return 1 + tail.size();
+  }
+}
+
+const s: { 
+  head: number, tail: List<number>,
+  size: () => number  
+} = new NonEmptyList(42);
+\end{lstlisting}
```
Comment (TimvdLippe) [inline report/sections/3-typescript.tex:84]: You compiled all the snippets right?

File: report/sections/3-typescript.tex [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+}
+const a: typeof NonEmptyList = NonEmptyList
+const xs: NonEmptyList<number> = new a(42)
+\end{lstlisting}
+
+From this it appears that classes live in both the value namespace as well as the type namespace,
+because an instance of \texttt{NonEmptyList} ($xs$) can be created from a runtime value ($a$) on the last line.
+In reality, $a$ represents the constructor for \texttt{NonEmptyList} rather than the class itself.
+We can therefore conclude that class constructors are first class values just like class methods.
+The type of a constructor must different from the type of a function because the former can only be called with the \texttt{new} keyword.
```
Comment (TimvdLippe) [inline report/sections/3-typescript.tex]: Must be

File: report/sections/3-typescript.tex [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+The \texttt{typeof} keyword in the definition of $a$ is used in this case to signal that $a$ is a constructor.
+Since constructors do not live inside the class scope, they have no impact type checking when handling instances of a class.
```
Comment (TimvdLippe) [inline report/sections/3-typescript.tex]: Impact on

File: report/sections/3-typescript.tex [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+\bigskip
+Inheritance is also handled in a straightforward manner.
+When checking whether an instance of a class can be assigned to an object type, the compiler
+simply takes all class members and all members of any parent class into account.
```
Comment (TimvdLippe) [inline report/sections/3-typescript.tex]: Add here that the fields of the value should be a superset of the fields specified in the object type

File: report/sections/4-syntax-definition.tex [hunk 11]
```diff
@@ -30,29 +31,52 @@ \section{Syntax definition}
...
+ModuleItem.ModuleStatement = 
+  StatementListItem
+\end{lstlisting}
+\caption{Definition of \textit{ModuleItem} in SDF3}
+\label{fig:direct-production-sdf}
+\end{figure}
+
+
+% \begin{figure*}
+%   \begin{minipage}[t]{.8\columnwidth}
```
Comment (TimvdLippe) [inline report/sections/4-syntax-definition.tex:61]: Why did you remove this one?

File: report/sections/5-type-checking.tex [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+A general solution to the name binding problem is provided by a framework of scope graphs.
+A scope graph consists of scopes, declarations and references to names in a program.
+Constraints can be generated to construct such graphs, as well as to resolve paths from declarations to references in the graph.
+Since the problem of type checking is very similar to that of name binding, NaBL2 can also be used to implement a type checker.
+
+\subsection{Structural type checking}
+
+In general, NaBL2 is well-suited for statements and expressions that are common in programming languages, 
+such as the shown \textit{IfStatement}.
+During this project, our research focussed on exploring to what extend NaBL2 supports structural type checking and,
```
Comment (TimvdLippe) [inline report/sections/5-type-checking.tex]: To what extent

File: report/sections/5-type-checking.tex [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+
+In general, NaBL2 is well-suited for statements and expressions that are common in programming languages, 
+such as the shown \textit{IfStatement}.
+During this project, our research focussed on exploring to what extend NaBL2 supports structural type checking and, 
+in the case that it does not, what would be required to support it.
+Our syntax definition for TypeScript includes several language constructs that must participate in type checking.
+These constructs included: object declarations, variable assignments, interfaces, functions, function calls and primitives.
+Several test cases and examples were developed that tested various aspects of structural type checking with these constructs.
+
+To demonstrate our attempt at implementing structural types, we will use object assignment as example.
```
Comment (TimvdLippe) [inline report/sections/5-type-checking.tex:23]: As an example

File: report/sections/5-type-checking.tex [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+The code below shows a TypeScript program that should be rejected by the type checker 
+because \texttt{Person} is a structural subtype of \texttt{Developer}.
```
Comment (TimvdLippe) [inline report/sections/5-type-checking.tex:25]: Is not

Comment (TimvdLippe) [reply report/sections/5-type-checking.tex:25]: Or it is, but this sentence is rather confusing. Better to remove double negation

File: report/sections/5-type-checking.tex [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+  language: string
+}
+var john: Person = { name: 'John' }
+var devJohn: Developer = john
+\end{lstlisting}
+
+\subsection{Naive solution}
+Object types can be modeled cleanly using scope graphs. 
+The general approach is to define a type, \texttt{RECORD}, that wraps a \texttt{scope}.
+The wrapped scope should contain the declatations of each field in the object type.
```
Comment (TimvdLippe) [inline report/sections/5-type-checking.tex]: Declarations

File: report/sections/5-type-checking.tex [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+    Value{name} <- s,
+    Value{name} : ty !.
+\end{lstlisting}
+\caption{Naive solution to object type checking.}
+\label{fig:wrong-solution}
+\end{figure*}
+
+While this solution is straightforward, there are two problems that prevent it from being useful.
+First, a pattern match on \texttt{ type@Id(\_) } is needed to dertermine that the left-hand side is indeed an object type. 
+This means that separate constraint generation rules are needed for checking non-object types, leading to code duplication.
```
Comment (TimvdLippe) [inline report/sections/5-type-checking.tex]: And if the id actually references an non-object type as an alias

File: report/sections/5-type-checking.tex [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+\subsection{Extensible subtype relations}
+The two problems illustrated above can be solved by making the NaBL2 subtype check user extensible.
+Normally, checking compatibility of two types in NaBL2 is done by generating constraints for equality or subtype checks using \texttt{==} or \texttt{<?}.
+In an ideal scenario, we would be able to use the \texttt{<?} operator to generate constraints for structural subtyping.
+However, this would require the operator to perform different checks based on the types of its arguments.
+For non-object types like strings and numbers, a simple non-recursive check is sufficient.
+Structural types require a recursive check. 
+As explained in section \ref{sec:structural-typing}, checking whether two object types in an equi-recursive system 
+(like with TypeScript) are subtypes requires recursively unfolding both types and checking their equality in the fixed point.
+Allowing users of NaBL2 to overload the \texttt{<?} operator is requires constraint generation and solving to be interleaved.
```
Comment (TimvdLippe) [inline report/sections/5-type-checking.tex]: This sentence has some grammar issues

File: report/sections/5-type-checking.tex [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+    Value{name} : ty!.
+\end{lstlisting}
+\caption{General solution when given the possibility of user-defined subtype checks.}
+\label{fig:cool-solution}
+\end{figure*}
+In the next section, the syntactical details of defining custom subtype relations are discussed.
+
+\subsection{NaBL2 syntax proposal}
+To syntactically support the definition of custom subtype relations, 
+we propose an extension to the SDF of NaBL2 as shown in Figure \ref{fig:nabl-syntax-proposal-relation}.
```
Comment (TimvdLippe) [inline report/sections/5-type-checking.tex]: Syntax definition. "The sdf" is not nice to read

File: report/sections/5-type-checking.tex [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+\end{itemize}
+Note that this relation pattern supports one constraint, 
+since multiple constraint can easilly be composed into one.
+We also propose two new constraint constructs: \texttt{forall} and \texttt{exists} as shown in Figure \ref{fig:nabl-syntax-proposal-constraint}.
+These constructs can be used to check that a constraint holds over all declarations in a scope, 
+or that there exists a declaration in a scope that satisfies a constraint.
+These three new syntactical constructs can be used to define a subtype relation for the \texttt{RECORD} type,
+illustrated in Figure \ref{fig:nabl-syntax-proposal-usage}.
+The \texttt{forall} will generate the constraint after the \texttt{=>} symbol for each declaration in the \texttt{one} scope.
+The \texttt{exists} construct ensures that \texttt{ty1 <? ty2} holds for at least one declatation in the \texttt{two} scope.
```
Comment (TimvdLippe) [inline report/sections/5-type-checking.tex]: Declaration

File: report/sections/5-type-checking.tex [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+The \texttt{sub : Type * Type} annotation on the seconds line determines that the name of the relation is \texttt{sub},
```
Comment (TimvdLippe) [inline report/sections/5-type-checking.tex]: Second line

File: report/sections/5-type-checking.tex [hunk 21]
```diff
@@ -0,0 +1,230 @@
...
+\end{lstlisting}
+\caption{Example usage of the new syntax constructs for structural types.}
+\label{fig:nabl-syntax-proposal-usage}
+\end{figure*}
+
+\subsection{Union and Intersection types}
+We will now demonstrate that with two additional, overloaded, versions of \texttt{forall} and \texttt{exists},
+NaBL2 will be powerful enough to implement union and Intersection types as described in Section \ref{sec:structural-typing-typescript}.
+Figure \ref{fig:union-impl} shows how this could be implemented.
+Notice that we user \texttt{forall} and \texttt{exists} in a manner similar to object types,
```
Comment (TimvdLippe) [inline report/sections/5-type-checking.tex]: We use

File: report/sections/5-type-checking.tex [hunk 22]
```diff
@@ -0,0 +1,230 @@
...
+\caption{Example usage of the new syntax constructs for structural types.}
+\label{fig:nabl-syntax-proposal-usage}
+\end{figure*}
+
+\subsection{Union and Intersection types}
+We will now demonstrate that with two additional, overloaded, versions of \texttt{forall} and \texttt{exists},
+NaBL2 will be powerful enough to implement union and Intersection types as described in Section \ref{sec:structural-typing-typescript}.
+Figure \ref{fig:union-impl} shows how this could be implemented.
+Notice that we user \texttt{forall} and \texttt{exists} in a manner similar to object types, 
+but rather operating on scopes, we now operate on lists of types.
```
Comment (TimvdLippe) [inline report/sections/5-type-checking.tex]: Rather than

**Review by TimvdLippe — APPROVED** [review]
Nice! Mostly some spelling mistakes.
